Unstructured-IO/pipeline-sec-filings
# chore: notebook with hideable cells

With this PR it is no longer necessary to keep a seperate version of the notebook `pipeline-section.ipynb` for demos.

#### Background
For the purposes of demos, markdown cells with detailed explanation are distracting, and it's desirable that code output is removed so it's clear the demo is "live". For these reasons we kept the notebook `pipeline-section.ipynb` as an example notebook, and a notebook derived from `pipeline-section.ipynb` for demos only.

This PR adds a javascript cell to the top of the notebook the output of which has a link that toggles the visibility of all cells with explanatory markdown. The javascript code also hides the input portion of its own cell since it is not relevant to the demo.

#### Explanation
The way this works is that the explanatory markdown cells have Jupyter tags. Cell tags can be viewed/changed via View -> Cell Toolbar -> Tags.
![image](https://user-images.githubusercontent.com/64741807/192630980-729a8b22-759b-4a8d-8b9d-5b9cdb41f382.png)

The javascript toggles the visibility of only those cells with the tag `explanatory_cell`, and hides the input of only those cells with the tag `execution_cell`.

HTML and javascript is active in a notebook when it is in the output of a cell. This notebook has been pre-executed so that the javascript is in the output of the first cell so it should execute as soon as the notebook loads -- however, Jupyter guards against this to prevent a notebook received from elsewhere from executing malicious code without any action from the notebook user. So on first loading the notebook the javascript in the output of the first cell is stripped automatically because the output is not *trusted*. The output of the script cell can be designated as trusted by either executing it locally, or going File -> Trust Notebook.
![image](https://user-images.githubusercontent.com/64741807/192631071-60ea4042-453b-4c15-ae35-dcefd98ac450.png)

This trust persists as long as any changes to the output are performed and recorded by the local Jupyter server. For more details see the [Jupyter security docs](https://jupyter-notebook.readthedocs.io/en/stable/security.html#the-problem).

#### Preparing the notebook for a demo
Once the output is trusted, the link appears in the output and can be clicked to hide explanatory markdown cells. Doing this in addition to Cell -> All Output -> Clear prepares the notebook for a demo.

Incidentally, clearing all cell output also clears the output of the script cell, erasing the link that allows you to toggle the visibility of the explanatory cells. It might be desirable for the link to be hidden for demos, but it can be made visible again by executing it again. Even though it is hidden, it can be selected by moving to the top of the notebook.
![image](https://user-images.githubusercontent.com/64741807/192631196-fe012d05-c03a-4a00-b05e-5abbca2ddc72.png)

(Sneaky hidden cell)

#### Testing
- Open the notebook `pipeline-section.ipynb`.
- Trust the notebook - the code for the first cell should disappear and the link should appear.
- Click the link and ensure explanatory markdown cells disappear (might have to click twice, I think this is a focus issue).
- Click the link again and ensure explanatory markdown cells reappear.
- Close and shutdown the notebook.
- Open the notebook again - code for the first cell should still be hidden and link should be visible.
- Run `make test-integration` to ensure notebook functionality is unchanged.
